NE-2104: desiredIstio: Enable GIE if InferencePool found

Miciah opened the pull request:
Configure the Istio control-plane to enable Gateway API Inference Extension (GIE) if the InferencePool CRD from GIE exists on the cluster.

Conditionally set one new environment variable in the Istio CR:

    ENABLE_GATEWAY_API_INFERENCE_EXTENSION

This change resolves [NE-2104](https://issues.redhat.com//browse/NE-2104).

aslakknutsen commented on pkg/operator/controller/gatewayclass/controller.go:
```diff
@@ -31,6 +32,10 @@ import (
+
+	// inferencepoolCrdName is the name of the InferencePool CRD from
+	// Gateway API Inference Extension.
+	inferencepoolCrdName = "inferencepools.inference.networking.k8s.io"
```
The group name is/has changed. 

v1.0.0/v1(not released)       inferencepools.inference.networking.k8s.io
v0.5.0/v1alpha2(released)  inferencepools.inference.networking.x-k8s.io

Miciah replied:
https://github.com/openshift/cluster-ingress-operator/compare/e48f4a8ebbec00bbc208c82f79e4cbfb513b25ec..585d99823e13c300b485e733211599f235347b15 changes the logic to check both API groups.

aslakknutsen commented:
@Miciah openshift-service-mesh/istio:master is == istio upstream (but slightly delayed). We back ported this latest change before it hit upstream. hehe

Miciah commented:
This time, [e2e-azure-operator failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1245/pull-ci-openshift-cluster-ingress-operator-master-e2e-azure-operator/1950372052498124800) because `TestAll/parallel/TestMTLSWithCRLs/certificate-distributes-its-own-crl` failed:
```text
    client_tls_test.go:1079: failed to get hostname of route "echo-pod-m9gzw"
```
It is possible the test needs a higher timeout.  In any case, more logging would be helpful.  

/test e2e-azure-operator

grzpiotrowski commented on pkg/operator/controller/gatewayclass/controller.go:
```diff
@@ -92,6 +100,20 @@ func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, er
 	}
 
+	// Watch for the InferencePool CRD to determine whether to enable
+	// Gateway API Inference Exception (GIE) on the Istio control-plane.
```
```suggestion
	// Gateway API Inference Extension (GIE) on the Istio control-plane.
```
nit: typo

Miciah replied:
Fixed in https://github.com/openshift/cluster-ingress-operator/compare/585d99823e13c300b485e733211599f235347b15..35c0143b7c5f4654544c7a834958ef8b371a60e1.

alebedev87 commented:
@Miciah : I forgot to ask. How did you test the change?

Miciah commented:
> @Miciah : I forgot to ask. How did you test the change?

I did some basic testing as follows, using a Cluster Bot cluster, to make sure that the watch, cache, and update logic worked as expected.  In one terminal, scale down CIO and run a local instance with the changes:
```shell
oc patch clusterversions/version --type=json --patch='[{"op":"add","path":"/spec/overrides","value":[{"kind":"Deployment","group":"apps","name":"ingress-operator","namespace":"openshift-ingress-operator","unmanaged":true}]}]'
oc -n openshift-ingress-operator scale deploy/ingress-operator --replicas=0
make run-local
```
In another terminal, enable Gateway API, and create and delete the CRD:
```shell
oc create -f -<<'EOF'                  
apiVersion: gateway.networking.k8s.io/v1beta1
kind: GatewayClass
metadata:
  name: openshift-default
spec:
  controllerName: openshift.io/gateway-controller/v1
EOF
git clone git@github.com/kubernetes-sigs:gateway-api-inference-extension ~/src/github.com/kubernetes-sigs/gateway-api-inference-extension
oc create -f ~/src/github.com/kubernetes-sigs/gateway-api-inference-extension/config/crd/bases/inference.networking.k8s.io_inferencepools.yaml
oc delete -f ~/src/github.com/kubernetes-sigs/gateway-api-inference-extension/config/crd/bases/inference.networking.k8s.io_inferencepools.yaml
```
CIO logged the updates to the Istio CR with the expected diff as the CRD was created and deleted (I did this a few times):
```diff
  v1.IstioSpec{                                 
        ... // 2 identical fields               
        Profile:   "",                          
        Namespace: "openshift-ingress",         
        Values: &v1.Values{                     
                Global: &v1.GlobalConfig{       
                        ... // 40 identical fields
                        IpFamilyPolicy:  nil,   
                        Waypoint:        nil,   
-                       TrustBundleName: nil,   
+                       TrustBundleName: &"openshift-gw-ca-root-cert",
                },                              
                Pilot: &v1.PilotConfig{         
                        ... // 13 identical fields
                        PodLabels: nil,         
                        ConfigMap: nil,         
                        Env: map[string]string{ 
                                ... // 9 identical entries
                                "PILOT_GATEWAY_API_DEFAULT_GATEWAYCLASS_NAME": "openshift-default",
                                "PILOT_MULTI_NETWORK_DISCOVER_GATEWAY_API":    "false",
-                               "SUPPORT_GATEWAY_API_INFERENCE_EXTENSION":     "true",
                        },                      
                        Affinity:        nil,   
                        RollingMaxSurge: nil,   
                        ... // 22 identical fields
                },                              
                Telemetry:              nil,    
                SidecarInjectorWebhook: &{EnableNamespacesByDefault: &false},
                ... // 9 identical fields       
        },                                      
  }                                             
```
and
```diff
  v1.IstioSpec{                                 
        ... // 2 identical fields               
        Profile:   "",                          
        Namespace: "openshift-ingress",         
        Values: &v1.Values{                     
                Global: &v1.GlobalConfig{       
                        ... // 40 identical fields
                        IpFamilyPolicy:  nil,   
                        Waypoint:        nil,   
-                       TrustBundleName: nil,   
+                       TrustBundleName: &"openshift-gw-ca-root-cert",
                },                              
                Pilot: &v1.PilotConfig{         
                        ... // 13 identical fields
                        PodLabels: nil,         
                        ConfigMap: nil,         
                        Env: map[string]string{ 
                                ... // 9 identical entries
                                "PILOT_GATEWAY_API_DEFAULT_GATEWAYCLASS_NAME": "openshift-default",
                                "PILOT_MULTI_NETWORK_DISCOVER_GATEWAY_API":    "false",
-                               "SUPPORT_GATEWAY_API_INFERENCE_EXTENSION":     "true",
                        },                      
                        Affinity:        nil,   
                        RollingMaxSurge: nil,   
                        ... // 22 identical fields
                },                              
                Telemetry:              nil,    
                SidecarInjectorWebhook: &{EnableNamespacesByDefault: &false},
                ... // 9 identical fields       
        },                                      
  }                                             
```
I can also see the environment variable being added and removed by monitoring it using `oc get istios -o yaml -w`.  

Note that I tested an earlier version, which used the old environment variable name (`SUPPORT_GATEWAY_API_INFERENCE_EXTENSION` instead of `ENABLE_GATEWAY_API_INFERENCE_EXTENSION`).  Also, the diff shows the bogus `trustBundleName` field updates, which #1243 fixed.

candita commented:
/assign @Thealisyed 
/assign @grzpiotrowski

Miciah commented:
We should merge #1244 before cherry-picking #1245 to the release-4.19 branch in order to avoid merge conflicts.

Miciah commented:
#1244 is merged.

/cherry-pick release-4.19
